### PR TITLE
Support resolving keyed services from DI in RDF and RDG

### PR DESCRIPTION
**Pull request by captainsafia**
Addresses https://github.com/dotnet/aspnetcore/issues/49633

### CODE REVIEW

File: src/Http/Http.Extensions/gen/StaticRouteHandlerModel/EndpointParameter.cs [hunk 1]
```diff
@@ -139,6 +139,12 @@ private void ProcessEndpointParameterSource(Endpoint endpoint, ISymbol symbol, I
         {
             Source = EndpointParameterSource.Service;
         }
+        else if (attributes.TryGetAttribute(wellKnownTypes.Get(WellKnownType.Microsoft_Extensions_DependencyInjection_FromKeyedServicesAttribute), out var keyedServicesAttribute))
```
Comment (BrennanConroy) [inline src/Http/Http.Extensions/gen/StaticRouteHandlerModel/EndpointParameter.cs:142]: Does this work with derived types?

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/RequestDelegateCreationTests.KeyServices.cs [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+    }
+
+    [Theory]
+    [InlineData(10)]
+    [InlineData(false)]
+    [InlineData(12.3)]
+    public async Task SupportsSingleKeyedServiceWithPrimitiveKeyTypes(object key)
+    {
+        var source = $$"""
+app.MapGet("/", (HttpContext context, [FromKeyedServices({{key.ToString()?.ToLowerInvariant()}})] TestService arg) => context.Items["arg"] = arg);
```
Comment (BrennanConroy) [inline src/Http/Http.Extensions/test/RequestDelegateGenerator/RequestDelegateCreationTests.KeyServices.cs:49]: Does this work? The key is a `string` but the keyed service is added with `object`.

File: src/Http/Http.Extensions/gen/StaticRouteHandlerModel/EndpointParameter.cs (hunk 1 above)
Comment (captainsafia) [reply src/Http/Http.Extensions/gen/StaticRouteHandlerModel/EndpointParameter.cs:142]: Are you referring to if the user is using a derived version of `FromKeyedServices` attribute? If so, the answer is no but not sure that is a scenario we want to support?

File: src/Http/Http.Extensions/test/RequestDelegateGenerator/RequestDelegateCreationTests.KeyServices.cs (hunk 2 above)
Comment (captainsafia) [reply src/Http/Http.Extensions/test/RequestDelegateGenerator/RequestDelegateCreationTests.KeyServices.cs:49]: The key here is actually an object. The `ToString()?.ToLowerVariant()` here is to get it to emit the write code for the `false` value. The source code snippet ends up having `[FromKeyedServices(false)]`, `[FromKeyedServices(12.3)]`, etc.

Comment (BrennanConroy) [reply src/Http/Http.Extensions/test/RequestDelegateGenerator/RequestDelegateCreationTests.KeyServices.cs:49]: Ok, stringed code is odd looking 😃

**Comment by captainsafia** [comment]
Closing so I can cherry-pick and retarget to 8.0-rc1.

**Pull request closed without merging**